### PR TITLE
New Dynamic Param schema and boilerplate

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Initial setup for the new DynamicParameter system as discussed in [this doc](https://docs.google.com/document/d/1-QD8_7TZiq6CNRSMeSmyg0ChrSVfvxwt_e6eLqiqRDA/edit?usp=sharing)
* Created new folder in `shared` since this will need to be available to both firmware and software now that we will generate C constants.
* Created an example config showing the new schema for all parameters
    * Notable new additions / changes include
        * The ability to share configs
        * Optional parameters
        * Const parameters
        * Support for enums
        * Support for factories
* Created skeleton code for the new generation script (in python)
* Created a custom bazel rule for generating dynamic parameters
    * This is needed to support enums and factories, because we need to include relevant headers which we extract from bazel targets

Please let me know if anything doesn't make sense or needs more clarification / documentation. I expect the bazel rule may need some more, along with _why_ certain things are needed (like enum_targets).

### Testing Done

Manual testing. Nothing to test for python yet, and can't really unit test bazel rules.

### Resolved Issues

#1296 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/shared/parameter_v2/BUILD
Comment (MathewMacDougall) [inline src/shared/parameter_v2/BUILD]: I opted to give this rule conditional behavior based on the file extension it's generating, so that we can easily make 2 targets (one for CPP and one for C). Are people fine with this or would they rather have two separate rules entirely? (In either case I expect the Python stuff to stay the same).

File: src/shared/parameter_v2/config/example.yaml
Comment (MathewMacDougall) [inline src/shared/parameter_v2/config/example.yaml]: Technically we could generate strings for C, they would just be `const char*`, which implies the `const` attribute in the YAML.

I really don't think we will ever need string constants in C, but at the same time it's generally not good to have this one-off conditional behavior in schemas. Thoughts on how to make this more intuitive for those writing configs?

Edit: Use case for strings in C is IP addresses, so this is valid.

File: src/shared/parameter_v2/parameter.bzl
Comment (MathewMacDougall) [inline src/shared/parameter_v2/parameter.bzl]: Just a note, `direct_headers` is only available in `bazel >= 3.1.0`. Most people are probably on `3.0.0` right now (I was). It's available in `apt` so should just be a matter of getting people to upgrade / re-run the setup scripts.

**Review by MathewMacDougall — COMMENTED** [review]
Just bringing special attention to a few things.

File: src/shared/parameter_v2/config/example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/example.yaml]: Updated the string param

File: src/shared/parameter_v2/config/example.yaml
Comment (EvanMorcom) [inline src/shared/parameter_v2/config/example.yaml]: What does 'could be a float'. mean in this case? Like it might as well be a float or it doesn't really matter

File: src/shared/parameter_v2/config/example.yaml
Comment (EvanMorcom) [inline src/shared/parameter_v2/config/example.yaml]: Same as above

File: src/shared/parameter_v2/config/example.yaml
Comment (EvanMorcom) [inline src/shared/parameter_v2/config/example.yaml]: I don't think 'better named' has a dash in it

File: src/shared/parameter_v2/parameter.bzl
Comment (EvanMorcom) [inline src/shared/parameter_v2/parameter.bzl]: We're going to use .hpp files now? Seems strange since we've been using .h for cpp for ages

**Review by EvanMorcom — COMMENTED** [review]
:fore :100:

File: src/shared/parameter_v2/BUILD
Comment (jonathanlew) [reply src/shared/parameter_v2/BUILD]: Seems like you're proposing that only dynamic parameters cpp header files get the .hpp extension and c header files get the .h extension. I don't like this idea, because the rest of our codebase has .h for cpp header files, so I think it would be too easy to people to intend a cpp header file and forget the .hpp. In effect, we'd have a special exception for dynamic params. I would prefer the rules set up in such a way that we explicitly declare which type of param file we want (separate rules or as an explicit argument to the rule).

File: src/shared/parameter_v2/BUILD
Comment (jonathanlew) [inline src/shared/parameter_v2/BUILD:6]: I don't understand what this is used for

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Happy to see this coming along

File: src/shared/parameter_v2/config/example.yaml
Comment (garethellis0) [reply src/shared/parameter_v2/config/example.yaml]: Hmm, one potential thought is that we can't have `double`'s in firmware here.

File: src/shared/parameter_v2/config/example.yaml
Comment (garethellis0) [inline src/shared/parameter_v2/config/example.yaml]: Would all this have to be indented and under a parameter name? ie 

```
My_param:
  metadata:
    ....
BoolParam:
  ...
```

File: src/shared/parameter_v2/config/example.yaml
Comment (garethellis0) [reply src/shared/parameter_v2/config/example.yaml]: I think this makes more sense, there is a semantic difference to signed vs unsigned numbers, `double` vs `float` is really just about precision.

File: src/shared/parameter_v2/config/example.yaml
Comment (garethellis0) [inline src/shared/parameter_v2/config/example.yaml]: Overall, I really like this, assuming that we can reasonably implement it.

File: src/shared/parameter_v2/parameter.bzl
Comment (garethellis0) [reply src/shared/parameter_v2/parameter.bzl]: I agree, this seems odd.

**Review by garethellis0 — APPROVED** [review]
Overall this is quite clean, looking forwards to seeing it working.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Whoops, sorry, meant to request changes.

File: src/shared/parameter_v2/config/example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/example.yaml]: "Could be a float" was me being lazy and pointing out we could support other floating-point types like `float`. 

Gareth is also right. Easiest thing I can think of is to enforce that anything generated for C must be defined as a float, and fail generation if we're trying to put a double there. This is probably better than auto-converting `doubles` to `floats` if they're being generated for C. Thoughts?

File: src/shared/parameter_v2/config/example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/example.yaml]: Yea same response as above, just me being lazy. In theory we could use any integral type here: `int_16`, `uint_8` but I also dont' think we generally have a use-case for those, so may be better to allow only things explicitly listed in the `example` file.

Thoughts on generically accepting int types vs being very explicit about what's specifically allowed?

File: src/shared/parameter_v2/parameter.bzl
Comment (MathewMacDougall) [reply src/shared/parameter_v2/parameter.bzl]: I  will split this rule up as Jon suggested above so the file extension won't matter.

File: src/shared/parameter_v2/config/example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/example.yaml]: No the `---` denotes a separate "document" in the YAML file. Each document can be read entirely separately. I chose to do it this way because then we can have one small set of metadata for the file, and also can avoid indenting everything.

See the section on "Multiple Documents" [here](https://rollout.io/blog/yaml-tutorial-everything-you-need-get-started/)

File: src/shared/parameter_v2/config/example.yaml
Comment (garethellis0) [reply src/shared/parameter_v2/config/example.yaml]: Yeah, I think the easiest option is just to have a rule that says that no shared parameters may be `double`'s.

File: src/shared/parameter_v2/config/example.yaml
Comment (garethellis0) [reply src/shared/parameter_v2/config/example.yaml]: IMO we should be explicit because that allows to control the types used to not use "non-standard" types and give the user more informative error messages at generation time, rather then fail at compile time.

File: src/shared/parameter_v2/BUILD
Comment (MathewMacDougall) [reply src/shared/parameter_v2/BUILD:6]: This is an example of how we need to include any targets that define an enum, in order to support it natively in the DynamicParam system.

Eg. If we wanted to autogenerate a `Parameter` that used an enum type, such as `RefboxGameState`, we need to include the header that `RefboxGamestate` is defined in so that we can build the autogenerated parameter code.

Because it's difficult to specify filepaths to other packages in bazel (without doing things like `exporting` files which would add confusion / complexity), we use paths to the targets where the enum is defined. The custom rule is able to extract the headers used to build these targets, giving us the include paths to add to the autogenerated code.

Lmk if that makes sense.

File: src/shared/parameter_v2/config/example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/example.yaml]: yup I agree. Will update the schema

**Comment by MathewMacDougall** [comment]
Gareth, Evan, and I just had a big chat where we discovered the current planned layout of configs (each file is a config, no nesting within files) doesn't work. Will work on updating the proposal doc to properly describe the issues and new solution, so we can have a quick meeting this week (probably after Sunday).

All other changes are still relevant though including the example schema in this PR, so please keep reviewing that. This PR won't be blocked by this issue.

File: src/shared/parameter_v2/BUILD
Comment (jonathanlew) [reply src/shared/parameter_v2/BUILD:6]: If we're reading the enums to create the dynamic param, then I was confused about the naming, because `enum_target` seems to imply that it's an output of this rule. How about `enum_srcs` or `enum_deps`?

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/shared/parameter_v2/config/example.yaml
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config/example.yaml]: Oh thats interesting, a pretty clean way of handling this

**Review by akhilveeraghanta — COMMENTED** [review]
LGTM :fire: pending `hpp` comments

File: src/shared/parameter_v2/BUILD
Comment (MathewMacDougall) [reply src/shared/parameter_v2/BUILD:6]: `enum_deps` is a better idea

File: src/shared/parameter_v2/BUILD
Comment (MathewMacDougall) [reply src/shared/parameter_v2/BUILD]: Opted to split into separate rules (used macros to avoid duplicating most of the implementation).

File: src/shared/parameter_v2/config/example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/example.yaml]: fixed

File: src/shared/parameter_v2/parameter.bzl
Comment (MathewMacDougall) [reply src/shared/parameter_v2/parameter.bzl]: resolved by splitting into separate rules (macros in this case) for c and cpp. See above comment.

File: src/shared/parameter_v2/BUILD
Comment (MathewMacDougall) [reply src/shared/parameter_v2/BUILD:6]: renamed to `enum_deps`

File: src/shared/parameter_v2/config/example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/example.yaml]: As suggested in another comment, I've now explicitly listed all valid parameter types. I included a note that `doubles` will not be generated for C in the description.

File: src/shared/parameter_v2/config/example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/example.yaml]: Updated the example to explicitly list all valid parameter types.

File: src/shared/parameter_v2/config/example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/example.yaml]: This is now changed based on the updated proposal in the dynamic param slack channel.

File: src/shared/parameter_v2/config/example.yaml
Comment (MathewMacDougall) [inline src/shared/parameter_v2/config/example.yaml]: Thoughts on this?

**Review by MathewMacDougall — COMMENTED** [review]
Addressed comments and slightly changed the example schema to match the new config structure as outlined in the updated proposal.

**Review by jonathanlew — DISMISSED** [review]
Looks clean! :+1:

**Review by EvanMorcom — DISMISSED** [review]
LGTM
🎸  Code be jamming 🎹

**Comment by MathewMacDougall** [comment]
PR has been updated according to the latest revisions to the proposal doc. Main changes are:
* Include paths will follow bazel's style (:MyConfig.yaml for configs in the same directory, otherwise an absolute path from the root dir of the configs must be provided)
* Files and directories will be named CamelCase only for configs (since this is a deviation from our regular convention)
* A config file "is" a config, ie. the name of the file is the name of the config, and parameters can go directoy at the top-level

**Comment by MathewMacDougall** [comment]
Updated the PR again based on the latest (and hopefully last) revision of the design doc.

File: src/shared/parameter_v2/config/examples/Example.yaml
Comment (garethellis0) [inline src/shared/parameter_v2/config/examples/Example.yaml]: Would be helpful to see the equivalent C/C++ `struct` corresponding to this.

Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/examples/Example.yaml]: Fair enough, but do you think that's super useful as documentation here? It is technically an implementation detail so we should be somewhat cautious about exposing too much.

Do you mean just see how a user would access the included config?

ie. `config->example->example_included_config_with_absolute_path->foo_bool`?

Comment (garethellis0) [reply src/shared/parameter_v2/config/examples/Example.yaml]: Yeah, technically an implementation detail, but I think we made a lot of changes to the top level config so that it would neatly translate to C/C++, so I think seeing that structure here would help give me confidence that we've come up something that actually does sanely translate.

File: src/shared/parameter_v2/config/examples/Example.yaml
Comment (akhilveeraghanta) [inline src/shared/parameter_v2/config/examples/Example.yaml]: I think we should put includes at the top like protobuf, and just have the files as a list
Since we say a file is a config, its redundant to name it again imo
~~~
include:
- foo.yaml
- test.yaml
~~~

Comment (garethellis0) [reply src/shared/parameter_v2/config/examples/Example.yaml]: That does create some trickyness if we have two configs with the same name at different parts of the tree. For example:
```
A
  B.yaml
C
  B.yaml
D
   B.yaml
   E.yaml
```
and say in `E.yaml` we have:
```
include:
- //A:B.yaml
- //C:B.yaml
- :B.yaml
```
Butttttt since we only ever want there to be a single instance of a config, I think it's totally fine to just disallow naming configs the same thing to prevent confusion.

Comment (jonathanlew) [reply src/shared/parameter_v2/config/examples/Example.yaml]: How would we enforce disallowing naming configs the same thing?

File: src/shared/parameter_v2/config/examples/Example.yaml
Comment (jonathanlew) [inline src/shared/parameter_v2/config/examples/Example.yaml]: What happens when we change values in the yaml? Do we have to flash firmware again? @akhilveeraghanta had an idea about sending a proto to the robots to update dynamic params, so we can avoid flashing firmware just to update dynamic params. In which case, dynamic params don't need to be consts.

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Got a question about how dyanamic params will work on firmware, but otherwise looks good!

File: src/shared/parameter_v2/config/examples/Example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/examples/Example.yaml]: I can throw generation-time errors if there are files in the config tree with the same name.

@garethellis0 are you fine with @akhilveeraghanta 's suggestion then if we disallow duplicate file names? We do lose the ability to "name" the included configs, but that's not a huge deal since config files _should_ be named reasonably (just goes back to our in-person discussion about making it as hard as possible to write something invalid, and giving very clear, local, error messages).

File: src/shared/parameter_v2/config/examples/Example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/examples/Example.yaml]: There still may be the distinction between "constants" and "dynamic params", even for firmware.

ie. we may want most values to be compiled-in, and only allow some to be changed. We don't necessarily want to rely on wifi/radio communication to update some possibly outdated set of params on the robots every time they boot up IMO

For now in the current proposal this will require re-flashing firmware to get updated values.

File: src/shared/parameter_v2/config/examples/Example.yaml
Comment (garethellis0) [reply src/shared/parameter_v2/config/examples/Example.yaml]: In terms of generating `C` structs, would we change the name from `CamelCase` to `snake_case`? Or would we just name the files with `snake_case` and then create the types by changing to `CamelCase`?

File: src/shared/parameter_v2/config/examples/Example.yaml
Comment (garethellis0) [reply src/shared/parameter_v2/config/examples/Example.yaml]: I think that constants are actually extremely useful for performance reasons, as there are some things (for example, robot mass) that are used very often, and we don't want to have to worry about spending the compute to make sure they're constantly updated. Obviously this would only impact a very small handful of cases, but I think it's still a valid consideration.

File: src/shared/parameter_v2/config/examples/Example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/examples/Example.yaml]: Ah yea I forgot about different naming conventions for different outputs. I originally wanted to avoid mutating names (eg. `snake_case` to `CamelCase`) during generation because I thought it would be misleading, but given that we have different standards for different outputs I think that's actually fine. It can be part of the spec for each language.

I'll hopefully get example generated code up soon which will help clarify this too.

@garethellis0 please confirm if you're ok with the
```
include:
- foo.yaml
- test.yaml
```

if includes, or would rather keep a version similar to what we have currently that allows includes to be named.

Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/examples/Example.yaml]: Talked to @garethellis0 , he agrees with akhil's suggestion to list includes at the top.

File: src/shared/parameter_v2/config/examples/Example.yaml
Comment (jonathanlew) [reply src/shared/parameter_v2/config/examples/Example.yaml]: OK so we'd want to track versions of these dynamic param constants in terms of fw hash?

Comment (garethellis0) [reply src/shared/parameter_v2/config/examples/Example.yaml]: Yeah, I think we'd need to generate a hash of all the constants, and then compare that with whatever is in firmware to make sure everything matches up (and warn if not). It's definitely a bit of a pain, so this could very reasonably be a slightly longer term thing.

Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/examples/Example.yaml]: this would also depend how we manage our `firmware_hash` system anyway? ie. to check if the firmware on the robots matches the version running with `ai`. I imagined this may be a hash of the built target rather than all the inputs, which would implicitly contain the firmware constants.

Comment (garethellis0) [reply src/shared/parameter_v2/config/examples/Example.yaml]: Yeah, if those constants were compiled in then you'd be correct.

**Review by jonathanlew — DISMISSED** [review]
LGTM pending other reviewers' comments

File: src/shared/parameter_v2/config/examples/example.yaml
Comment (MathewMacDougall) [inline src/shared/parameter_v2/config/examples/example.yaml:34]: Ok so in order for the YAML to be parsed successfully, we need to "terminate" this `include` list somehow. This is because you can't have a "top-level list" with a dictionary list like this.

The two options are:
* terminate the list with `---` (what we have now)
* Put all parameters under their own dictionary

Eg.
```
include:
  - "foo.yaml"
parameters:
  - bool:
      name: example_bool_param
      value: true
      description: >
        Can be true or false
  - int:
      name: example_bool_param
      min: 0
      max: 5
      value: 3
      description: >
        Can be any integer value in the range [min, max]
```

Do people have a preference for one or the other?

**Comment by MathewMacDougall** [comment]
Made #1447 for future work creating example generated code. Updated the PR based on previous feedback
* file naming convention is back to `snake_case.yaml` since we will need to modify the strings on a per-language basis to suite our style guide in the generated code.
* `includes` are now a single list in the config, and just the filename needs to be provided since filenames are unique
* Please see my comment above

**Review by jonathanlew — DISMISSED** [review]
LGTM pending CI

File: src/shared/parameter_v2/config/examples/example.yaml
Comment (MathewMacDougall) [reply src/shared/parameter_v2/config/examples/example.yaml:34]: @jonathanlew I don't know if :+1: means you like the first or second option

Comment (jonathanlew) [reply src/shared/parameter_v2/config/examples/example.yaml:34]: I mean to say that's a good point you bring up. I personally have no strong feelings either way, so I approved the PR

Comment (akhilveeraghanta) [reply src/shared/parameter_v2/config/examples/example.yaml:34]: I vouch for `---` obvious separation from includes to rest of the schema

**Review by akhilveeraghanta — DISMISSED** [review]
LGTM :fire:

File: src/shared/parameter_v2/config/examples/example.yaml
Comment (garethellis0) [inline src/shared/parameter_v2/config/examples/example.yaml]: Why not just have a single floating point type that's represented as a `float` on the `C` side and a double on the `C++` side?

File: src/shared/parameter_v2/config/examples/example.yaml
Comment (garethellis0) [inline src/shared/parameter_v2/config/examples/example.yaml:84]: Should this be `MAKE_ENUM` instead?

**Review by akhilveeraghanta — DISMISSED** [review]
LGTM :satisfied_matthew:

**Review by jonathanlew — COMMENTED** [review]
Lgtm pending ci

**Review by garethellis0 — DISMISSED** [review]
Great work pushing this through, thanks for all the patience with :infinity: revisions

LGTM :+1: :tada:

**Comment by MathewMacDougall** [comment]
Fixed CI, need re-approval

**Review by jonathanlew — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]
:shipit:

**Pull request merged**
